helix-core/src/main/java/org/apache/helix/messaging/handling/HelixStateTransitionHandler.java

```diff
@@ -463,6 +442,36 @@ public void onError(Exception e, ErrorCode code, ErrorType type) {
 
   }
 
+  // Verify the fromState and current state of the stateModel.
+  private void validateStaleMessage (boolean isPreCheck) throws Exception {
+    String fromState = _message.getFromState();
+    String toState = _message.getToState();
+    String partitionName = _message.getPartitionName();
+
+    // state in _currentStateDelta uses current state from state model. It has the
+    // most up-to-date. current state. In case currentState in stateModel is null,
+    // partition is in initial state and we using it as current state.
+    // Defined in HelixStateMachineEngine.
+    String state = _currentStateDelta.getState(partitionName);
+
+    if (toState.equalsIgnoreCase(state)) {
+      // To state equals current state, we can just ignore the message
+      throw new HelixDuplicatedStateTransitionException(String
+          .format("Partition %s current state is same as toState (%s->%s) from message.",
+              partitionName, fromState, toState));
+    } else if (!isPreCheck && fromState != null && !fromState.equals("*") && !fromState
```

Comment:
Why we need this precheck flag? Two checks in onMessage and HelixTask execution validation logic should be same

Reply:
TFTR. 
Originally in HelixStateTransitionHandler, handling for these two cases are different.
For the first case when toState == currentState, we just treat the task as succeeded and finish task. There is no difference between doing the no-op when executing and bypass the task at scheduling phase.

In the second case when currentState != fromState, executor will set the task as failed and report state transition error. Ir will also update ZK.

Having same logic causes a test fail. Controller keeps sending many 'INIT->RUNNING' message when currentState is TASK_ERROR if we don't report task error. Currently task error is reported in HelixStateTransitionHandler.

Reply:
For the test you talked about, I think we may need to understand why we keep sending message INIT -> RUNNING. Because once the message has been DROPPED, controller should refresh current state and get the TASK_ERROR state. Then we should have some message like TASK_ERROR -> INIT or something, if we miss it, we should fix the logic to make it more robust. And I am OK to not fix it in this PR. But if that's the case, let's propose another PR for that. Otherwise, this is not a complete scenario. For regular resource management, it is not a complete filtering logic.

Reply:
I had an offline sync with Ali. He has submit a PR to fix this issue (#1390). I changed my PR accordingly. 
Currently my PR contains his patch.  Will rebase after his PR merged.